### DISCUSSION TITLE
Login, Register, and Admin Actions

### DISCUSSION THREAD

**Discussion by mango009** [zitadel/zitadel#4323] [Q&A]
This is more of a request than a question.

We have evaluated several open source and IAM services, we narrowed down to **KeyCloak**, **Ory Kratos**, and **Zitadel**. We really like **Zitadel** for it's clean architecture and extensibility. But we have to stick with **KeyCloak** for one **Zitadel** show-stopper. We need the ability of being able to listen to `login`, `register`, and `admin` events as available in [**KeyCloak**](https://www.keycloak.org/docs/14.0/server_admin/#login-events). 

**Zitadel** allows similar capability via `action` but it supports only for external authentication flow. Also we find having to implement action using Javascript a little inconvenient (and maybe insecure). Maybe a cleaner way is for intended listeners to register Webhook callback. **KeyCloak** achieves this by using `SPI`. 

We hope **Zitadel** can provide compatible events as those provided by **KeyCloak**soon and we can come back to it.

**Comment by adlerhurst** [comment]
Hi @mango009 thank you very much for your thoughts. We will discuss your points and give you feedback asap :)

**Comment by mffap** [comment]
> **Zitadel** allows similar capability via `action` but it supports only for external authentication flow. Also we find having to implement action using Javascript a little inconvenient (and maybe insecure). Maybe a cleaner way is for intended listeners to register Webhook callback. **KeyCloak** achieves this by using `SPI`.

At the moment the feature is still early in development. We're currently working towards [Phase 1 of this issue](https://github.com/zitadel/zitadel/issues/4265) in the next 2 weeks. You could add the required triggers to the issue.

You can also see that we intend to ship some QoL improvements. What are your concerns regarding security of Javascript?

There's an issue for gauging interest in webhook support: https://github.com/zitadel/zitadel/issues/13
For me this seems however as a more specific case than the generalized Actions capabilities. In terms of convenience, however this seems definitely easier.

FYI: We wanted to achieve extensibility the platform without having to compile and release ZITADEL itself, that's why we disregarded the SPI framework approach of Keycloak.


Edit: Only Phase 1 will land in the next 2 weeks.

**Reply by mango009** [reply]
Looking forward to the update on the event extension!

I'm not in favor of SPI framework especially if the code has to be done in Javascript. I guess this code has to be executed inside the Zitadel runtime and there would be lots of opportunities for malicious code execution.

Webhook is a much cleaner way to go as this decouples the security concern away from Zitadel. Also, we do not need to put our business logic in the javascript code exposed to Zitadel. Hope you guys can pick up momentum on this as I see it almost got dropped from your issue tracking.

**Reply by mffap** [reply]
> Looking forward to the update on the event extension!

👍

> I'm not in favor of SPI framework especially if the code has to be done in Javascript. I guess this code has to be executed inside the Zitadel runtime and there would be lots of opportunities for malicious code execution.

Maybe, if you're not yet on our Discord. There was also a couple of Discussions, eg. [here](https://discord.com/channels/927474939156643850/927866013545025566/1002632931581382686), about the execution. The runtime itself is isolated, but obviously some binding to the system exists.
We're currently in the process of a routine pentest, where especially this feature is also in focus. We will publish the report on our Blog once mitigation is done.

> Webhook is a much cleaner way to go as this decouples the security concern away from Zitadel. Also, we do not need to put our business logic in the javascript code exposed to Zitadel. Hope you guys can pick up momentum on this as I see it almost got dropped from your issue tracking.

You mean the stale bot? We'll someone found that to be a good idea :) No it's not dropped. Please comment on that issue or show a reaction. That helps us a lot to prioritize. Thanks.
